### PR TITLE
[import] Check for and use existing module imports for type annotations

### PR DESCRIPTION
**Pull request by ezragoss**
If we want to annotate type `Name` from `mod` and there exists the import statement:
`import mod`
then we will annotate `mod.Name` rather than add the import statement
`from mod import Name`

### CODE REVIEW

**Comment by chadrik** [comment]
Can you rebase this against master, and squash commits.  I added mypy to the pre-commit checks, and I see a number of typing problems here.

File: typewriter/fixes/base.py [hunk 1]
```diff
@@ -722,7 +725,8 @@ def type_updater(self, match):
         if '.' not in word and ':' not in word:
             # Assume it's either builtin or from `typing`
             if word in typing_all:
-                self.add_import('typing', word)
+                if not does_tree_import('typing', word, node):
```
Comment (chadrik) [inline typewriter/fixes/base.py]: As far as I can tell there's no guarantee here that `node` will not be `None`.   Would `does_tree_import` be ok with that?  Should we just enforce that `node` is never `None`?

File: typewriter/fixes/fixer_utils.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    Returns
+    -------------
+    Optional[str]
+        The valid name of the type based on current imports
+    """
+    root = find_root(node)
+
+    if does_tree_import(package, name, root):
+        # name is being imported directly
+        return name
```
Comment (chadrik) [inline typewriter/fixes/fixer_utils.py:34]: Can you clarify in your comments what style of import this is. e.g.  `from <package> import <name>`.  Same for the one below.

File: typewriter/fixes/fixer_utils.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    Parameters
+    -------------
+    package : str
+    name : str
+    node : Node
+
+    Returns
+    -------------
+    Optional[str]
+        The valid name of the type based on current imports
```
Comment (chadrik) [inline typewriter/fixes/fixer_utils.py]: "Returns None if an import for `name` already exists"

File: typewriter/fixes/base.py [hunk 4]
```diff
@@ -733,5 +737,14 @@ def type_updater(self, match):
         else:
             mod, name = word.rsplit('.', 1)
             to_import = name
-        self.add_import(mod, to_import)
+
+        if node is not None:
+            # Get the typename we need to use to be valid for a current import statement
+            result = type_by_import_stmt(mod, to_import, node)
+
+        if result is not None:
```
Comment (chadrik) [inline typewriter/fixes/base.py:765]: seems like `result` may be undefined if `node` is `None`

File: typewriter/fixes/base.py (hunk 1 above)
Comment (ezragoss) [reply typewriter/fixes/base.py]: Yeah let's just enforce it as a non-optional argument. At this point it is crucial to the functionality of adding typing import statements

File: typewriter/fixes/fixer_utils.py (hunk 2 above)
Comment (ezragoss) [reply typewriter/fixes/fixer_utils.py:34]: done

File: typewriter/fixes/fixer_utils.py (hunk 3 above)
Comment (ezragoss) [reply typewriter/fixes/fixer_utils.py]: It would be "Returns None if no import for `name` already exists" but added!

File: typewriter/fixes/base.py (hunk 4 above)
Comment (ezragoss) [reply typewriter/fixes/base.py:765]: solved by enforcing node as not None and removing superfluous check

File: typewriter/fixes/base.py [hunk 5]
```diff
@@ -741,5 +745,13 @@ def type_updater(self, match):
...
-        self.add_import(mod, to_import)
+
+        # Get the typename we need to use to be valid for a current import statement
+        result = type_by_import_stmt(mod, to_import, node)
+
+        if result is not None:
+            # There exists a current import statement this type is valid for
+            return result
+
+        self.add_import(mod, to_import)  # There exists no import statement for this type, add one
```
Comment (chadrik) [inline typewriter/fixes/base.py]: add comment on its own line before this, rather than trailing after it.

File: typewriter/fixes/fix_annotate_docs.py [hunk 6]
```diff
@@ -152,14 +152,16 @@ def parse_docstring(self, docstring, line):
...
+        # type: (Match, Node) -> str
         # Replace `pkg.mod.SomeClass` with `SomeClass`
         # and remember to import it.
         word = match.group()
         # Assume it's either builtin or from `typing`
         if word in typing_all:
-            self.add_import('typing', word)
+            if not does_tree_import('typing', word, node):
+                # Check whether import exists already
+                self.add_import('typing', word)
```
Comment (chadrik) [inline typewriter/fixes/fix_annotate_docs.py]: I see this pattern repeated twice:

```python
        if word in typing_all:
            if not does_tree_import('typing', word, node):
                # Check whether import exists already
                self.add_import('typing', word)
```

I think it would be nice to wrap it up into a `add_typing_import(word, node)` method.

File: typewriter/fixes/fixer_utils.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        # Mod is being imported in the form of
+        # import <package>
+        return '.'.join((package, name))
+
+    return  # return None to signal there is not yet an import statement
+
+
+def create_import(package, name, node):
+    """
+    Create import statement for type <name> from <package>
```
Comment (chadrik) [inline typewriter/fixes/fixer_utils.py]: what form of import does this create?

Can we reword this as:

```
Create import statement of the form `from <package> import <name>`
```

or whatever is correct

Comment (ezragoss) [reply typewriter/fixes/fixer_utils.py]: done

File: typewriter/fixes/fix_annotate_docs.py (hunk 6 above)
Comment (ezragoss) [reply typewriter/fixes/fix_annotate_docs.py]: i added `touch_typing_import` because I think that's more accurate but if you think it's misleading just let me know

File: typewriter/fixes/base.py (hunk 5 above)
Comment (ezragoss) [reply typewriter/fixes/base.py]: done

**Comment by chadrik** [comment]
Rebase against master and run `pre-commit run -a` again.  It will generate type annotations from your docstrings, which reveals a few remaining issues with mypy. 

Also, it's not enforced anywhere yet, but this project has a 99 character max line length.  Can you adjust your changes accordingly, please.

File: typewriter/fixes/fixer_utils.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def create_import(package, name, node):
+    """
+    Create import statement of the form `from <package> import <name>`
+
+    Parameters
+    -------------
+    package : str
+    name : str
```
Comment (chadrik) [inline typewriter/fixes/fixer_utils.py:52]: It would be good to clarify here (and in other docstrings with the name argument) that `name` refers to a type.

**Comment by chadrik** [comment]
> Rebase against master and run `pre-commit run -a` again. It will generate type annotations from your docstrings, which reveals a few remaining issues with mypy.

Heads up: you'll noticed that the typewriter pre-commit task always fails with "files were modified by this hook".  This is because the typewriter code does not properly find items imported using parentheses (`from foo import (x, y, z)`) and so adds new import statements to the code, which are then removed by the subsequent `isort` task, which correctly detects that they are redundant. 

We will fix this issue in a followup pull request.

File: typewriter/fixes/fixer_utils.py (hunk 8 above)
Comment (ezragoss) [reply typewriter/fixes/fixer_utils.py:52]: done

**Comment by ezragoss** [comment]
^ Noted
^^ Fixed the mypy issues. It's been rebased and type annotations have been generated.
